### ISSUE TITLE
The view takes more space than it needs to

### ISSUE THREAD

**Issue by Konrad-Wszelaki** [deinstapel/cursive-aligned-view#6]
I'm working on a project where I found this crate to be useful, but the first time I tried using it I found that it worked in terms of aligning the views, but at the same time it took literally all of the space that was available instead of just what it needs to take.
Here's how it looked the first time around:
![bad_layout](https://user-images.githubusercontent.com/58418997/120606089-15139f00-c44f-11eb-89cd-768e1a26c686.png)
And here's how it was supposed to look:
![good_layout](https://user-images.githubusercontent.com/58418997/120606121-1f359d80-c44f-11eb-8253-0d1e27890a47.png)

As you can see, I did make it do what I wanted it to do. I did it by changing your required_size() function so it returns 'self.last_size' instead of 'constraint'. Doing that fixes the issue for me completely, so I've decided to leave that info here in case you find it useful (not making a pull request since I'm not sure whether that's a good fix in general or just in my case).

**Comment by fin-ger** [comment]
Hi, thanks for pointing this out!

I assembled a quick example which shows the issue.

<details>
<summary>Example code</summary>

```rust
use cursive::view::{Resizable, SizeConstraint};
use cursive::views::{DummyView, Panel, LinearLayout, TextView, Button};
use cursive::{Cursive, CursiveExt};
use cursive_aligned_view::Alignable;

fn main() {
    let mut siv = Cursive::default();
    
    let left = Panel::new(DummyView).title("Left panel");
    let bottom = Panel::new(DummyView).title("Bottom panel");
    let right_top = Panel::new(DummyView).title("Right top panel");
    let right_bottom = Panel::new(
        LinearLayout::vertical()
            .child(TextView::new("Press this button to quit"))
            .child(Button::new("Quit", |s| s.quit()))
            .align_center()
            .resized(SizeConstraint::Free, SizeConstraint::Free)
        ).title("Right bottom panel");
        
    let layout = LinearLayout::vertical().child(
        LinearLayout::horizontal().child(left).child(
            LinearLayout::vertical().child(right_top).child(right_bottom)
        )
    ).child(bottom);

    siv.add_fullscreen_layer(layout);
    siv.run();
}
```

</details>

I extended this with a `Free` size constraint for width and height which clearly shows that our layouting misbehaves. The two relevant size constraints for this issue are

```
Free - No constraint imposed, the child view's response is used.
Full - Tries to take all available space, no matter what the child needs.
```

which implies that `Free` should not have the same behavior as `Full` and that `Free` will only allocate what is actually needed for the child to be displayed properly. I think we are violating this.

@jwuensche I think this change will break our end2end snapshot tests. We might also include tests for different size constraints set via the `Resizable` trait. What do you think? Should we make this exhaustive and create tests for all usage scenarios being

 - 9 (top left, top center, top right, right center, bottom right, bottom center, bottom left, left center, center) alignments settable via trait and view methods (18 in total)
 - 5 size constraints (Free, Full, Fixed, AtMost, AtLeast)
 - So we have 5*18=90 tests in total

What do you think? Is this too much?

**Comment by fia0** [comment]
> which implies that Free should not have the same behavior as Full and that Free will only allocate what is
> actually needed for the child to be displayed properly. I think we are violating this.

I think this really just leaves it open for the purpose of the view to decide what is best. But I agree that in our use-case we should not make the assumption for the user and return here the maximum size. If we do this the usage of the lib is much more annoying with often occurring `.resized(AtMost(some_size))` or other restrictions to fit in interfaces. Instead the response to from the child is probably best to be used.

> @jwuensche I think this change will break our end2end snapshot tests. We might also include tests for 
> different size constraints set via the Resizable trait. What do you think? Should we make this exhaustive 
> and create tests for all usage scenarios being
> ...
> What do you think? Is this too much?

We could construct test cases fewer in numbers. Basically we have three possible categories for the size allocation, we get less, more or equal to the amount we requested. How this constraints came about doesn't really make a difference to us, as we are simply informed by cursive in the draw call about the given size.
1. we receive less space
  This _should_ be one case (?) of the view being squished, taking the alignment part out of the picture and simply rendering the child with the space available.
2. we receive equal space
  Similar to before but the view receives the requested space.
3. we receive more space
  Where we actually need to align parts of the view. These should be the most cases, they are testable with either `Full`, `Fixed` or `AtLeast` constraints, which one we choose does not really make a difference to test for our library, as we do not handle these constraints.

> alignments settable via trait and view methods (18 in total)

Why test the alignments in separate snapshots. It would be sufficient to ensure in tests if the properties are actually set via trait or view method correctly. And later on test one of the verified variants as a snapshot test. This would lead to only `5*9=45` snapshots (without restrictions from before) and `5*9+18=63` tests in total.

**Comment by fia0** [comment]
So coming back to this after a few days. I have fixed the issue with the view constraint sizes and added and updated the tests which we need as mentioned before. All the changes are available in`v0.5.0`.
The changes to the behavior are the same to the ones you made locally, so everything should work ootb for you. If there are any problems please let us know! Thanks for raising this issue anyway and helping us improve this lib :smile:
